platform/mbed_error.h

```diff
@@ -23,13 +22,793 @@
+                      ERROR_CODE_##error_name = MBED_CUSTOM_ERROR_BASE + error_code,  \
+                      ERROR_##error_name = MAKE_MBED_ERROR(ERROR_TYPE_CUSTOM, MODULE_UNKNOWN, ERROR_CODE_##error_name)
+
 
+/**
+ * Macro for setting a system error. This macro will log the error, prints the error report and return to the caller. Its a wrapper for calling set_error API.
+ * @param  error_status     mbed_error_status_t status to be set(See mbed_error_status_t enum above for available error status values).
+ * @param  error_msg        The error message to be printed out to STDIO/Serial.
+ * @param  error_value      Value associated with the error status. This would depend on error code/error scenario.
+ *
+ * @code
+ * 
+ * SET_ERROR( ERROR_INVALID_SIZE, "MyDriver: Invalid size in read", size_val )
+ *
+ * @endcode
+ * @note The macro calls set_error API with filename and line number info without caller explicitly passing them.
+ *        Since this macro is a wrapper for set_error API callers should process the return value from this macro which is the return value from calling set_error API.
+ *
+ */
+#ifdef MBED_CONF_ERROR_FILENAME_CAPTURE_ENABLED
+    #define SET_WARNING( error_status, error_msg, error_value )             set_warning( error_status, (const char *)error_msg, (uint32_t)error_value, (const char *)MBED_FILENAME, __LINE__ )
+#else
+    #define SET_WARNING( error_status, error_msg, error_value )             set_warning( error_status, (const char *)error_msg, (uint32_t)error_value, NULL, 0 )
+#endif
+
+/**
+ * Macro for setting a fatal system error. This macro will log the error, prints the error report and halts the system. Its a wrapper for calling set_error API
+ * @param  error_status     mbed_error_status_t status to be set(See mbed_error_status_t enum above for available error status values).
+ * @param  error_msg        The error message to be printed out to STDIO/Serial.
+ * @param  error_value      Value associated with the error status. This would depend on error code/error scenario.
+ * @return                  0 or ERROR_SUCCESS.
+ *                          ERROR_INVALID_ARGUMENT if called with invalid error status/codes
+ *
+ * @code
+ * 
+ * SET_ERROR( ERROR_MUTEX_LOCK_FAILED, "MyDriver: Can't lock driver Mutex", &my_mutex )
+ *
+ * @endcode
+ * @note The macro calls set_error API with filename and line number info without caller explicitly passing them.
+*        Since this macro is a wrapper for set_error API callers should process the return value from this macro which is the return value from calling set_error API. 
+ *
+ */
+#ifdef MBED_CONF_ERROR_FILENAME_CAPTURE_ENABLED
+    #define SET_ERROR( error_status, error_msg, error_value )       set_error( error_status, (const char *)error_msg, (uint32_t)error_value, (const char *)MBED_FILENAME, __LINE__ )
+#else
+    #define SET_ERROR( error_status, error_msg, error_value )       set_error( error_status, (const char *)error_msg, (uint32_t)error_value, NULL, 0 )
```

Comment:
SET_ERROR is still a confusing name for a function that halts the processor

Reply:
What would be a good name? I dont like calling it just error because it sounds like we are declaring/defining something. I want that to be clear that it will be acted upon.

Reply:
Sorry, I'm coming in the middle of the review a bit so might have missed some of the nuanced/offline discussions. The current use of `error(...)` halts execution. To me calling this macro `ERROR` makes more sense as `SET_ERROR` implies setting the status of an error code. At least to me `ERROR` implies it will be acted on.